### PR TITLE
ZK Service Status Tracking

### PR DESCRIPTION
**Pull request by stkem**
@andrewconner @eishay

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/FortyTwoGlobal.scala [hunk 1]
```diff
@@ -63,10 +63,15 @@ abstract class FortyTwoGlobal(val mode: Mode.Mode)
     
     val amazonInstanceInfo = injector.instance[AmazonInstanceInfo]
     log.info(s"Amazon up! $amazonInstanceInfo")
-    injector.instance[ServiceDiscovery].register()
+    val serviceDiscovery = injector.instance[ServiceDiscovery]
+    serviceDiscovery.register()
+    serviceDiscovery.startSelfCheck
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/FortyTwoGlobal.scala]: if its an action method (i.e. you can't conceptually replace with with `var`) please use parentheses as in `serviceDiscovery.startSelfCheck()`

File: kifi-backend/modules/common/app/com/keepit/FortyTwoGlobal.scala [hunk 2]
```diff
@@ -91,6 +99,7 @@ abstract class FortyTwoGlobal(val mode: Mode.Mode)
         e.printStackTrace
         log.error(errorMessage, e)
     }
+    serviceDiscovery.unRegister()
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/FortyTwoGlobal.scala]: maybe have unRegister on `try ... finally` ?

File: kifi-backend/modules/common/app/com/keepit/FortyTwoGlobal.scala [hunk 3]
```diff
@@ -63,10 +63,15 @@ abstract class FortyTwoGlobal(val mode: Mode.Mode)
...
-    injector.instance[ServiceDiscovery].register()
+    val serviceDiscovery = injector.instance[ServiceDiscovery]
+    serviceDiscovery.register()
+    serviceDiscovery.startSelfCheck
+
   }
 
   override def onError(request: RequestHeader, ex: Throwable): Result = {
+    val serviceDiscovery = injector.instance[ServiceDiscovery]
+    serviceDiscovery.changeStatus(ServiceStatus.SICK)
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/FortyTwoGlobal.scala:74]: is there a way to un-sick instances if an admin thinks the exception is not too bad?
I fear of random exceptions suddenly making a cluster of 10 to have 9 sick instances and one poor instance that gets all the load (until a fix will be on its way).

File: kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceCluster.scala [hunk 4]
```diff
@@ -36,22 +36,28 @@ class ServiceCluster(val serviceType: ServiceType) extends Logging {
   override def toString(): String = s"""Service Cluster of $serviceType:
     instances.toString"""
 
-  //using round robin
-  def nextService(): Option[ServiceInstance] = {
-    val list = routingList
+  //using round robin, fall back on also using SICK instances if there are no healthy (i.e. UP) ones left
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceCluster.scala]: cool!
maybe we should have a watermark threshold of "if more then 50% of the instances are sick, keep on going until help comes" ?

File: kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceDiscovery.scala [hunk 5]
```diff
@@ -88,21 +90,36 @@ class ServiceDiscoveryImpl @Inject() (
     log.info(s"registered clusters: $clusters, my service is $myServiceType")
     val myCluster = clusters(myServiceType)
     val instanceInfo = amazonInstanceInfoProvider.get
-    myNode = Some(zk.createNode(myCluster.serviceNodeMaster, Json.toJson(instanceInfo).toString, EPHEMERAL_SEQUENTIAL))
-    myCluster.register(myNode.get, instanceInfo)
+    val thisRemoteService = RemoteService(instanceInfo, ServiceStatus.STARTING, myServiceType)
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceDiscovery.scala:93]: nice!

File: kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceCluster.scala (hunk 4 above)
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceCluster.scala]: I'm not quite sure what you mean. If there are more than 50% sick instances, treat them as healthy?

File: kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceDiscovery.scala [hunk 6]
```diff
@@ -88,21 +90,36 @@ class ServiceDiscoveryImpl @Inject() (
...
+      thisServiceInstance.foreach{ serviceInstance =>
+        serviceInstance.remoteService.status = newStatus
+        zk.set(node, RemoteService.toJson(serviceInstance.remoteService))
+      }
+    }
+  }
+
+  def startSelfCheck(): Unit = future {
+    if(services.currentService.selfCheck) changeStatus(ServiceStatus.UP)
+    else changeStatus(ServiceStatus.DOWN)
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceDiscovery.scala]: how about changing the state to HEALTHCKECK_FAIL or something similar?
in some cases (emergency) we may say "yea, we know healthcheck fail but its a known issue, we can't roll back the db patch that caused it and we must manually make sure the server is up until we'll get the fix in"

**Pull request merged**